DLPNO-CCSD(T) (#3293)
psi4/psi4: doc/sphinxman/source/dlpnoccsd_t.rst

such as crambin (636 atoms) [Riplinger:2013:134101]_ and insulin [Jiang:2024:082502]_ have been performed
with this algorithm. This algorithm was originally implemented in the ORCA software package, but it
is now available in |PSIfour|!

For a more comprehensive overview on local correlation and the DLPNO-CCSD(T) algorithm, the reader is referred
to the :ref:`DLPNO-MP2 <sec:dlpnomp2>` documentation, and the published work of Jiang et al. describing the
implementation of DLPNO-CCSD(T) in |Psifour| [Jiang:2024:082502]_.

An example input file for a DLPNO-CCSD computation is::
   
   memory 2 GB

   molecule h2o {
   0 1
   O
   H 1 1.0
   H 1 1.0 2 104.5
   symmetry c1
   }
   
   set basis cc-pvdz
   set scf_type df
   set freeze_core True
   set pno_convergence tight
   
   energy('dlpno-ccsd')

An example input file for a DLPNO-CCSD(T) computation is::

   memory 2 GB

   molecule h2o {
   0 1
   O
   H 1 1.0
   H 1 1.0 2 104.5
   symmetry c1
   }
   
   set basis cc-pvdz
   set scf_type df
   set freeze_core True
   set pno_convergence tight
   
   energy('dlpno-ccsd(t)') # dlpno-ccsd(t0) for the semicanonical (T0) computation

> 

Author: Done!

  and computations involving DLPNO-CCSD/(T) are encouraged to be performed on large molecules 
  as a more accurate alternative to DFT.

* For most computations, |dlpno__pno_convergence| ``TIGHT`` is recommended, especially 
  those involving non-covalent interactions. For larger systems where ``TIGHT`` is too expensive, 
  ``NORMAL`` for |dlpno__pno_convergence| while setting |dlpno_t_cut_pairs| to ``1.0e-5`` is recommended. 
  This has been shown to yield errors on the order of kJ/mol for non-covalent interactions [Jiang:2024:082502]_.

* For the triples computation, the semicanonical (T0) approach given in [Riplinger:2013:134101]_ is NOT 
  recommended. As opposed to ORCA, the default computation for a ``dlpno-ccsd(t)`` call is |PSIfour| is 
  the iterative (T) computation given in [Guo:2018:011101]_ (i.e. DLPNO-CCSD(T1) in ORCA). The semicanonical (T0) 
  implementation of triples can be requested by the user through the ``energy('dlpno-ccsd(t0)')`` call.
  (NOTE: In ORCA, a DLPNO-CCSD(T) computation defaults to the semicanonical (T0) computation.)

* Due to differences in implementation and working equations, absolute DLPNO-CCSD(T) energies in |PSIfour| will not

>   exactly match ORCA, as |PSIfours| |dlpno__pno_convergence| at a given level is designed to be more accuarate than

Reviewer: accurate

* Based on user allocated memory, disk/core storage for various integrals and tensors
  for DLPNO-CCSD/(T) are automatically determined. There is no need to toggle with the disk/core
  options for the average user.

* In DLPNO methods, it is recommended to freeze core orbitals (by setting |globals__freeze_core|
  to ``True``), since core excitations are known to be more sensitive to PNO truncations than
  valence truncations. If a non-frozen core computation is requested, all PNOs corresponding to core-core
  or core-virtual pairs have cutoffs scaled by ``T_CUT_PNO_CORE_SCALE`` (default ``1.0e-2``).

* Note that DLPNO does not yet have molecular point group symmetry implemented and will run in C1 symmetry.

* At this time, DLPNO-CCSD/(T) is only available for closed-shell RHF computations.

Computation Size Limits
-----------------------

* Since DLPNO-CCSD(T) is linear-scaling, with access to sufficient computing resources, a DLPNO-CCSD(T) computation 
  can be possible with any system. In fact, for larger systems, Hartree-Fock becomes the bottleneck (not the coupled-cluster)!

>   Below, we tabulate the projected limits (number of atoms) of DLPNO-CCSD(T) across different access to hardware. All these limits 

Reviewer 2: ```suggestion
  Below, we tabulate the roughly projected limits (number of atoms) of DLPNO-CCSD(T) across different access to hardware. All these limits 
```

Really nice all the rules of thumb you've added.

Key Differences with DLPNO-CCSD(T) in ORCA
------------------------------------------

We note that while the DLPNO-CCSD(T) algorithm in Psi4 is heavily inspired by the original method
proposed by Neese and coworkers in ORCA [Riplinger:2013:034106]_ [Riplinger:2013:134101]_ [Riplinger:2016:024109]_ 
[Guo:2018:011101]_, the DLPNO-CCSD(T) code in |PSIfour| is fundamentally represents a different algorithm than the one
in ORCA. In other words, the DLPNO-CCSD(T) code in |PSIfour| satisfies the same specifications as the DLPNO-CCSD(T)
algorithm in ORCA, but represents a different implementation. The best analogy is chicken sandwiches at different fast food chains
(e.g. Chick-fil-A vs. Popeyes). The food item represents a specification (i.e. fried chicken breast served between buttered, 
toasted buns), but the implementation is different (different oils, cooking temperatures, breading techniques, seasonings, etc).
Both algorithms represent linear-scaling CCSD(T) algorithms solved in the the local pair natural orbital basis, with convergence
to canonical CCSD(T) results as the local tolerances are tightened. However, the manner in which the PNO spaces are truncated as

> well as how the CCSD equations are solved are different. Notable differences in implementation between the two algorithms are

Reviewer 2: ```suggestion
While the DLPNO-CCSD(T) formulation in |PSIfour| is heavily inspired by the original method
proposed by Neese and coworkers in ORCA [Riplinger:2013:034106]_ [Riplinger:2013:134101]_ [Riplinger:2016:024109]_ 
[Guo:2018:011101]_, in places |PSIfour| employs different algorithms. 
Both represent linear-scaling CCSD(T) algorithms solved in the the local pair natural orbital basis, with convergence
to canonical CCSD(T) results as the local tolerances are tightened. However, the manner in which the PNO spaces are truncated as
well as how the CCSD equations are solved are different. Notable differences in implementation between the two algorithms are
```

Author: Ah, I understand the international audience would not understand my Chicken Sandwich reference. To be fair, I was sleep deprived when I wrote some of the docs :)

highlighted below:

* The most notable difference is that the DLPNO-CCSD equations in |PSIfour| utilizes T1-dressed Hamiltonian and Fock matrix elements,
  which significantly simplifies the number of working equations as well as the number of mathematical operations involved in solving
  the CCSD residual equations. Because of the reduced number of intermediates that are required, we find that the DLPNO-CCSD code in 
  |PSIfour| uses less RAM (around a factor of 2 to 3) than the corresponding implementation in ORCA, at a given |dlpno__pno_convergence|.
  The runtimes are expected to be similar.

* |PSIfour| often recovers slightly more CCSD correlation energy than ORCA, since additional PNO cutoffs are utilized in |PSIfour| 

>   compared to ORCA, with |dlpno__t_cut_trace| and |dlpno__t_cut_energy| cutoffs used in addition to the normal |dlpno__t_cut_pno|. 

Reviewer 2: ```suggestion
* |PSIfour| often recovers slightly more CCSD correlation energy than ORCA due to additional PNO cutoffs  
  |dlpno__t_cut_trace| and |dlpno__t_cut_energy| used in addition to the normal |dlpno__t_cut_pno|. 
```

Is it definitely general that p4 recovers more?

   
   set basis cc-pvdz
   set scf_type df
   set freeze_core True
   set pno_convergence tight
   
   energy('dlpno-ccsd')

An example input file for a DLPNO-CCSD(T) computation is::

   memory 2 GB

   molecule h2o {
   0 1
   O
   H 1 1.0
   H 1 1.0 2 104.5
   symmetry c1
   }
   
   set basis cc-pvdz
   set scf_type df
   set freeze_core True
   set pno_convergence tight
   
   energy('dlpno-ccsd(t)') # dlpno-ccsd(t0) for the semicanonical (T0) computation

PNO Convergence Settings
------------------------

Here we present a table of the PNO convergence settings, paramaters, and recommended use cases. Most of these parameters and
settings are similar to what is found in ORCA, with two added parameters (|dlpno__t_cut_trace| and |dlpno__t_cut_energy|) to
increase the robustness of the PNO space. These added parameters truncate by percent recovery of the total occupation number,
as well as the percentage energy recovery of the PNOs compared to the non-truncated basis.

.. _`table:pno_convergence`:

.. table:: PNO convergence levels given in |Psifour|

   +-------------+------------+-------------+--------------+----------+-----------+-------------+---------------------------+

>    | Convergence | T_CUT_PNO  | T_CUT_TRACE | T_CUT_ENERGY | T_CUT_DO | T_CUT_MKN | T_CUT_PAIRS | Recommended Applications  |

Reviewer 2: ```suggestion
   | |dlpno__pno_convergence| | T_CUT_PNO  | T_CUT_TRACE | T_CUT_ENERGY | T_CUT_DO | T_CUT_MKN | T_CUT_PAIRS | Recommended Applications  |
```
rest of the table will need formatting adjustment to compensate

  and computations involving DLPNO-CCSD/(T) are encouraged to be performed on large molecules 
  as a more accurate alternative to DFT.

* For most computations, |dlpno__pno_convergence| ``TIGHT`` is recommended, especially 
  those involving non-covalent interactions. For larger systems where ``TIGHT`` is too expensive, 
  ``NORMAL`` for |dlpno__pno_convergence| while setting |dlpno_t_cut_pairs| to ``1.0e-5`` is recommended. 
  This has been shown to yield errors on the order of kJ/mol for non-covalent interactions [Jiang:2024:082502]_.

* Based on user allocated memory, disk/core storage for various integrals and tensors
  for DLPNO-CCSD/(T) are automatically determined. There is no need to toggle with the disk/core
  options for the average user.

* In DLPNO methods, it is recommended to freeze core orbitals (by setting |globals__freeze_core|
  to ``True``), since core excitations are known to be more sensitive to PNO truncations than
  valence truncations. If a non-frozen core computation is requested, all PNOs corresponding to core-core
  or core-virtual pairs have cutoffs scaled by |dlpno_t_cut_pno_core_scale| (default ``1.0e-2``).

> * Please specify ``symmetry c1`` in the user input file. DLPNO does not yet have molecular point group symmetry implemented.

Author: Yes, correct! Will make that change
